common/rpm.go

```diff
@@ -245,27 +246,35 @@ func (rpm *ringpopMonitorImpl) FindHostForAddr(service string, addr string) (*Ho
 
 // FindHostForKey finds and returns the host responsible for handling the given key
 func (rpm *ringpopMonitorImpl) FindHostForKey(service string, key string) (*HostInfo, error) {
-	// Note: we don't have Lookup(key, predicate..)
-	// so just get all reachable hosts return the first instance
+
+	// get list of hosts that for the given service
 	members, err := rpm.GetHosts(service)
 	if err != nil || members == nil || len(members) == 0 {
 		return nil, ErrUnknownService
 	}
 
-	// We need to make sure we always return the same node
-	// So sort members by IP and return the lowest.
+	// this function should be consistent in picking and returning the
+	// same node for a given key on a specific 'set' of hosts; so we sort
+	// all the hosts by their ip-addresses and use a hash of the key to
+	// pick the host from that list.
 	var ipAddrs []string
-	tempMap := make(map[string]int, len(members))
-	for k, member := range members {
+	hostInfoMap := make(map[string]*HostInfo, len(members))
```

Comment:
I think we can avoid the temp map creation and sorting for every call. 
How about we keep the membershipInfo.asList() sorted, when we create it (see rpm.refresh()) ?
Then all this method has to do is - 

members := GetHosts()
return members[hash%len(members)].Addr